File: src/main/java/failchat/handlers/IgnoreFilter.java
package failchat.handlers;

import failchat.core.Configurator;
import failchat.core.Message;
import failchat.core.MessageFilter;
import failchat.core.Source;

import java.util.HashSet;
import java.util.Set;
import java.util.logging.Logger;
import java.util.regex.Pattern;

/**
 * Фильтрует сообщения от пользователей в игнор-листе
 * Баны хранятся в формате nickname#source
 * */
public class IgnoreFilter implements MessageFilter {
    private static final Logger logger = Logger.getLogger(IgnoreFilter.class.getName());
    private Set<String> ignoreSet = new HashSet<>();
    private Pattern banFormat;

    public IgnoreFilter() {
        StringBuilder sb = new StringBuilder();
        sb.append('(');
        for (Source source : Source.values()) {
            sb.append(source.getLowerCased()).append('|');
        }
        sb.deleteCharAt(sb.length() - 1).append(')');
        banFormat = Pattern.compile(".+#" + sb.toString());

        reloadIgnoreList();
    }

    @Override
    public boolean filterMessage(Message message) {
        return ignoreSet.contains(message.getAuthor() + "#" + message.getSource().getLowerCased());
    }

    public void ignore(String user) {
        ignoreSet.add(user);
        Configurator.config.setProperty("ignore", ignoreSet.toArray());
        logger.fine("User ignored: " + user);
    }

    public void reloadIgnoreList() {
        ignoreSet.clear();
        Configurator.config.getList("ignore").forEach((bannedUser) -> {
            if (banFormat.matcher((String) bannedUser).find())
                ignoreSet.add((String) bannedUser);
        });
    }
}

File: src/main/kotlin/failchat/core/chat/handlers/IgnoreFilter.kt
package failchat.core.chat.handlers

import failchat.core.Origin
import failchat.core.chat.ChatMessage
import failchat.core.chat.MessageFilter
import failchat.util.debug
import failchat.util.value
import org.apache.commons.configuration2.Configuration
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import java.util.concurrent.atomic.AtomicReference
import java.util.regex.Pattern

/**
 * Фильтрует сообщения от пользователей в игнор-листе.
 * Баны хранятся в формате authorId#origin.
 */
class IgnoreFilter(private val config: Configuration) : MessageFilter<ChatMessage> {

    private companion object {
        val log: Logger = LoggerFactory.getLogger(IgnoreFilter::class.java)
    }

    private val banFormat: Pattern

    private var atomicIgnoreSet: AtomicReference<Set<String>> = AtomicReference(emptySet())

    init {
        val originsPattern = Origin.values()
                .map { it.name }
                .joinToString(separator = "|", prefix = "(", postfix = ")")

        banFormat = Pattern.compile(".+#" + originsPattern)

        reloadConfig()
    }

    override fun filterMessage(message: ChatMessage): Boolean {
        val ignoreMessage = atomicIgnoreSet.value.contains(message.author.id + "#" + message.origin.name)
        if (ignoreMessage) log.debug { "Message filtered by ignore filter: $message" }
        return ignoreMessage
    }

    fun reloadConfig() {
        atomicIgnoreSet.value = config.getList("ignore")
                .map { it as String }
                .filter { banFormat.matcher(it).find() }
                .toSet()
        log.debug("IgnoreFilter reloaded a config")
    }

}
